src/codegen.cpp

```
@@ -492,22 +492,25 @@ bool CodeGen::isLowLevelTy(llvm::Value *val) const {
 
-void CodeGen::checkLowLevelTypeMix(llvm::Value *lhs, llvm::Value *rhs, const std::string &op) {
-    std::string lhsName = getLowLevelTypeName(lhs);
-    std::string rhsName = getLowLevelTypeName(rhs);
+void CodeGen::checkLowLevelTypeMix(llvm::Value *lhs, llvm::Value *rhs, const std::string &op,
+                                    const std::string &lhsHint, const std::string &rhsHint) {
+    // Fall back to AST suffix hints for constants that lack metadata (#311, #595)
+    const std::string &lhsRef = getLowLevelTypeName(lhs);
+    const std::string &rhsRef = getLowLevelTypeName(rhs);
+    const std::string &lhsName = !lhsRef.empty() ? lhsRef : lhsHint;
+    const std::string &rhsName = !rhsRef.empty() ? rhsRef : rhsHint;
     bool lhsLL = !lhsName.empty() || isLowLevelTy(lhs->getType());
```

Clang (14.0.6)

[warning] variable 'lhsLL' is not initialized

(cppcoreguidelines-init-variables)